lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,770 @@
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-widths"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Constraint Expressions
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct Expr;
+} // namespace
+
+/// Allow rvalue refs to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T &e) {
+  e.print(os);
+  return os;
+}
+
+/// Allow pointers to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T *e) {
+  return os << *e;
+}
+
+// Allow expression subclasses to be hashed.
+namespace mlir {
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::hash_code hash_value(const T &x) {
+  return x.hash_value();
+}
+} // namespace mlir
+
+namespace {
+
+/// An expression on the right-hand side of a constraint.
+struct Expr {
+  enum Kind {
+    KindRoot,
+    KindVar,
+    KindKnown,
+    KindAdd,
+    KindMax,
+    KindMin,
+  };
+  Kind kind;
+  llvm::Optional<int32_t> solution = {};
+
+  /// Print a human-readable representation of this expr.
+  void print(llvm::raw_ostream &os) const;
+
+  // Iterators over the child expressions.
+  typedef Expr *const *iterator;
+  iterator begin() const;
+  iterator end() const;
+
+protected:
+  Expr(Kind kind) : kind(kind) {}
+  llvm::hash_code hash_value() const { return llvm::hash_value(kind); }
+};
+
+/// Helper class to CRTP-derive common functions.
+template <class DerivedT, Expr::Kind DerivedKind>
+struct ExprBase : public Expr {
+  ExprBase() : Expr(DerivedKind) {}
+  static bool classof(const Expr *e) { return e->kind == DerivedKind; }
+  bool operator==(const Expr &other) const {
+    if (auto otherSame = dyn_cast<DerivedT>(other))
+      return *this == otherSame;
+    return false;
+  }
+  iterator begin() const { return nullptr; }
+  iterator end() const { return nullptr; }
+};
+
+/// A free variable.
+struct RootExpr : public ExprBase<RootExpr, Expr::KindRoot> {
+  void print(llvm::raw_ostream &os) const { os << "root"; }
+  iterator begin() const { return &exprs[0]; }
+  iterator end() const { return &exprs[0] + exprs.size(); }
+  std::vector<Expr *> exprs;
```

Comment:
Would SmallVector be reasonable here?

With either case you have the issue that this gives RootExpr a dtor.  Plz be careful that they don't enter a bump pointer allocator

Reply:
It looks like this is only used in one place.  Could/should the vector get moved out of RootExpr into the containing `ConstraintSolver`?

Reply:
The `std::is_trivially_destructible<T>` constraint on the `InternedAllocator` should help prevent that kind of use. But anyway you're right that the `RootExpr` is just there for the sake of regularity in the `GraphTraits`, and really doesn't need to hold the expression vector itself.

Reply:
It will be hard to pick a good size for `SmallVector` here, because of the width inference being a global task (needs to work across modules). I'd expect the number of expressions to be proportional to the input size, and not something more predictable/stable like size per module.

I'm happy to put a `SmallVector` there though if you have a better intuition regarding its size, though!

Reply:
Either way works for me, thanks!